Sources/ASTGen/ASTGen.swift

```diff
@@ -0,0 +1,253 @@
+import SwiftParser
+import SwiftSyntax
+
+import CASTBridging
+
+extension Array {
+  public func withBridgedArrayRef<T>(_ c: (BridgedArrayRef) -> T) -> T {
+    withUnsafeBytes { buf in
+      c(BridgedArrayRef(data: buf.baseAddress!, numElements: count))
+    }
+  }
+}
+
+extension UnsafePointer {
+  public var raw: UnsafeMutableRawPointer {
+    UnsafeMutableRawPointer(mutating: self)
+  }
+}
+
+struct ASTGenVisitor: SyntaxTransformVisitor {
+  let ctx: UnsafeMutableRawPointer
+  let base: UnsafePointer<CChar>
+
+  // TOOD: we need to be up updating this.
+  var declContext: UnsafeMutableRawPointer
+  
+  // TODO: this some how messes up the witness table when I uncomment it locally :/
+//  public func visit<T>(_ node: T?) -> [UnsafeMutableRawPointer]? {
+//    if let node = node { return visit(node) }
+//    return nil
+//  }
+  
+  @_disfavoredOverload
+  public func visit(_ node: SourceFileSyntax) -> UnsafeMutableRawPointer {
+    fatalError("Use other overload.")
+  }
+  
+  public func visit(_ node: SourceFileSyntax) -> [UnsafeMutableRawPointer] {
+    let loc = self.base.advanced(by: node.position.utf8Offset).raw
+    var out = [UnsafeMutableRawPointer]()
+
+    for element in node.statements {
+      let swiftASTNodes = visit(element)
+      if element.item.is(StmtSyntax.self) {
+        out.append(SwiftTopLevelCodeDecl_createStmt(ctx, declContext, loc, swiftASTNodes, loc))
+      } else if element.item.is(ExprSyntax.self) {
+        out.append(SwiftTopLevelCodeDecl_createExpr(ctx, declContext, loc, swiftASTNodes, loc))
+      } else {
+        assert(element.item.is(DeclSyntax.self))
+        out.append(swiftASTNodes)
+      }
+    }
+    
+    return out
+  }
+
+  public func visit(_ node: FunctionCallExprSyntax) -> UnsafeMutableRawPointer {
```

Comment:
I’m not really a fan that all these functions return `UnsafeMutableRawPointer`. I imagine it will be to easy to e.g. pass a `FunctionDecl` to a parameter that really expects an `Identifier` Can we do something more typed here? Like structs that wrap a `void *` or something like that?

Reply:
I was going to mention this on the C++ side, but since you have here - IMO it'd be good to define a bunch of opaque types (eg. `typedef void *swift_source_loc_t`) and use those instead of `void *` directly.

And another comment mentions labels - we could just add `swift_name` to the C functions, eg. `__attribute__((swift_name("FuncDecl_create(ctx:staticLoc:....)")))`